Resolve false negative for `used-before-assignment`  with TYPE_CHECKING guard

zenlyj opened the pull request:
<!--
Thank you for submitting a PR to pylint!

To ease the process of reviewing your PR, do make sure to complete the following boxes.

- [ ] Document your change, if it is a non-trivial one.
  - A maintainer might label the issue ``skip-news`` if the change does not need to be in the changelog.
  - Otherwise, create a news fragment with ``towncrier create <IssueNumber>.<type>`` which will be
    included in the changelog. ``<type>`` can be one of the types defined in `./towncrier.toml`.
    If necessary you can write details or offer examples on how the new change is supposed to work.
  - Generating the doc is done with ``tox -e docs``
- [ ] Relate your change to an issue in the tracker if such an issue exists (Refs #1234, Closes #1234)
- [ ] Write comprehensive commit messages and/or a good description of what the PR does.
- [ ] Keep the change small, separate the consensual changes from the opinionated one.
  Don't hesitate to open multiple PRs if the change requires it. If your review is so
  big it requires to actually plan and allocate time to review, it's more likely
  that it's going to go stale.
- [ ] If you used multiple emails or multiple names when contributing, add your mails
      and preferred name in ``script/.contributors_aliases.json``
-->

## Type of Changes

<!-- Leave the corresponding lines for the applicable type of change: -->

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
|    | :sparkles: New feature |
|    | :hammer: Refactoring   |
|    | :scroll: Docs          |

## Description

In the function responsible for filtering nodes defined under a `TYPE_CHECKING` block from consumption, usage scopes should only be updated when `used-before-assignment` is reported.

However, there is currently a misalignment between the `_report_unfound_name_definition()` and the filter function regarding when the error is raised. To resolve this, `_report_unfound_name_definition()` has been updated to return the check result, which is now passed to the filter function.

Additionally, the check has been extended to handle cases where a class is defined within an `if` block and used later.

<!-- If this PR references an issue without fixing it: -->

<!-- If this PR fixes an issue, use the following to automatically close when we merge: -->

Closes #8893

zenlyj commented:
setting this to draft first, as i would like to check this more thoroughly. but feel free to discuss / give feedback!

zenlyj commented:
@jacobtylerwalls can i get a review for this PR?

regarding the primer result, the sentry one seems legitimate. as an import is made under `TYPE_CHECKING`, then used for an `isinstance` check. the django ones should be reasonable reports as well? where a class is defined within an `if` block, but not within the associated `else` block, hence raising `possibly-used-before-assignment`

jacobtylerwalls commented on pylint/checkers/variables.py:
```diff
@@ -723,6 +723,8 @@ def _uncertain_nodes_if_tests(
                 name = other_node.name
             elif isinstance(other_node, (nodes.Import, nodes.ImportFrom)):
                 name = node.name
+            elif isinstance(other_node, nodes.ClassDef):
```
```suggestion
            elif isinstance(other_node, (nodes.FunctionDef, nodes.ClassDef)):
```

jacobtylerwalls commented on tests/functional/u/used/used_before_assignment_typing.py:
```diff
@@ -174,7 +174,7 @@ def defined_in_elif_branch(self) -> calendar.Calendar:  # [possibly-used-before-
     def defined_in_else_branch(self) -> urlopen:
         print(zoneinfo)  # [used-before-assignment]
         print(pprint())
```
```suggestion
        print(pprint())  # [used-before-assignment]
```

jacobtylerwalls reviewed (commented):
Thank you! Just a small request to handle one more case.

jacobtylerwalls commented:
Any idea about the new Django message for a nonlocal?

zenlyj commented:
seems that the message was raised due to a guarded function definition further down at line 1584. but from my understanding, this is a false positive? it should not flag as `condition` is provided as a function parameter in the enclosing function `_deferredSkip()`

jacobtylerwalls commented:
I wonder if we need to explicitly handle nonlocals somewhere. If you don't have bandwidth then I guess we can back out your last change and punt this to a new issue so that we don't cause more false positives.

zenlyj commented:
perhaps we should drop the latest commit and track this case with a separate issue? so that we can unblock #9964.

you may assign the issue to me and i'll find time to work on it

jacobtylerwalls commented:
Very good--thank you!

nickdrozd commented:
There is also the problem of `unused-import` false negatives:

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from math import pi  # unused-import false negative
```

Is it possible to fix that along the way?

jacobtylerwalls commented:
> you may assign the issue to me and i'll find time to work on it

Again, thanks for this offer. I opened #10028

zenlyj commented:
> > you may assign the issue to me and i'll find time to work on it
> 
> Again, thanks for this offer. I opened #10028

hey no worries, i dropped the latest commit so this should be good to go.



> There is also the problem of `unused-import` false negatives:
> 
> ```python
> from typing import TYPE_CHECKING
> 
> if TYPE_CHECKING:
>     from math import pi  # unused-import false negative
> ```
> 
> Is it possible to fix that along the way?

sure thing, i added it to #10028

Pierre-Sassoulas reviewed (approved):
The sentry primer looks legit.
